### PR TITLE
Fix unintended CodeModel output ordering after Prechecker was added

### PR DESCRIPTION
**Pull request by daviwil**
Testing some things, will turn this into a real PR if I find a fix.

### CODE REVIEW

File: modelerfour/prenamer/prenamer.ts [hunk 1]
```diff
@@ -133,7 +130,7 @@ export class PreNamer {
 
     // choice
     const choiceSchemaNames = new Set<string>();
-    for (const schema of values(this.codeModel.schemas.choices).toArray().sort(schemaNameComparer)) {
```
Comment (daviwil) [inline modelerfour/prenamer/prenamer.ts:136]: Sorting schemas before deduplicating names causes changes in the output ordering.  I know you recommended that I do this @fearthecowboy, do you remember why?

**Comment by daviwil** [comment]
There's a remaining issue that I can't easily explain yet, will investigate on Monday.

**Comment by pakrym** [comment]
Got a an error that I don't understand when running the following input

```md
input-file:
  - https://github.com/Azure/azure-rest-api-specs/blob/49fc16354df7211f8392c56884a3437138317d1f/specification/search/data-plane/Azure.Search/preview/2019-05-06-preview/searchindex.json
  - https://github.com/Azure/azure-rest-api-specs/blob/49fc16354df7211f8392c56884a3437138317d1f/specification/search/data-plane/Azure.Search/preview/2019-05-06-preview/searchservice.json

directive:
- from: searchindex.json
  where: $.definitions.SearchError
  transform: $["x-ms-client-name"] = "IndexSearchError"
```

Error:
```
ERROR (PreCheck/DuplicateSchema): Duplicate Schema named SearchError -- properties.details.$ref: undefined => "#/components/schemas/schemas:477",x-ms-client-name: "IndexSearchError" => <none>
FATAL: Error: 1 errors occured -- cannot continue.
  Error: Plugin prechecker reported failure.
```

After removing the transform:
```

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named SearchError -- properties.details.$ref: undefined => "#/components/schemas/schemas:477"
```

**Comment by pakrym** [comment]
Also see some polymorphic classes entirely missing from yaml:

#### body-complex:
image

This is our set of options: 
```
use-extension:
  "@autorest/modelerfour": "http://tinyurl.com/yckwga6x"
modelerfour:
  always-create-content-type-parameter: true
  flatten-models: true
  flatten-payloads: true
  group-parameters: true
  lenient-model-deduplication: true
```

**Comment by iscai-msft** [comment]
I generated all of the test swaggers and everything looks good besides what @pakrym mentioned, I'm missing Cookiecuttershark.
I'll also include my m4 config. It's the following plus a bunch of naming config (let me know if you'd like to see those too, I'm not including it right now because I don't think it's relevant)

use-extension:
  "@autorest/modelerfour": "http://tinyurl.com/yckwga6x"

modelerfour:
  group-parameters: true
  flatten-models: true
  flatten-payloads: true
  resolve-schema-name-collisons: true
  always-create-content-type-parameter: true
  multiple-request-parameter-flattening: false

Thanks for all of your work on this @daviwil!

**Comment by daviwil** [comment]
@pakrym @iscai-msft - It appears this happens because the `fixUpSchemasThatUseAllOfInsteadOfJustRef` check in Prechecker removes any schemas like `cookiecuttershark` which contain nothing other than an `allOf` with a reference to another schema.  `cookiecuttershark` is defined exactly in that way:

```
    "cookiecuttershark": {
      "type":  "object",
      "allOf": [
        {
          "$ref": "#/definitions/shark"
        }
      ]
    },
```

@fearthecowboy Do you think the pruning of these ostensibly empty schemas might be a little too aggressive?  It might be a pattern in use that could be hard to get rid off in the short term.  Should we just raise the warning and not drop the schema?

@lmazuel @MiYanni do you know what `cookiecuttershark` is meant to test in the `body-complex.json` spec?  Is there a known service spec that models a polymorphic type in this way which might require that we continue to support this type of schema (for now)?

**Comment by pakrym** [comment]
Seems like removing it loses polymorphic information (the type of the model)

**Comment by fearthecowboy** [comment]
`fixUpSchemasThatUseAllOfInsteadOfJustRef ` is intended to fix a terrible pattern that has been popping up recently.

Because OAI2/3 doesn't technically allow a `$ref` to have a description beside it, people have taken to using `allOf` and 'inheriting' the model they are trying to reference. (even tho' autorest has always allowed it). There is an OAI library for c# which spits them out like this, and tools are doing this because of that -- and we're seeing this in the specs repo. (swashbuckle to start with)

It's not technically invalid, but it really obscures the intent (which was to reference the other type, not inherit it.)

I went and tracked down the person who used this in the specs repo and he was just doing it by rote, since someone else did it. 

One could argue that the `cookiecuttershark` type isn't real because it's just inheriting the parent, without adding anything.

We have to be able to correct for that -- if we don't, we're going to get a ton of types generated that  end up having the same name (which is how I wandered across it in the first place).

I'd hate to flip the behavior behind a flag that because it's not going to go away, and people don't know what they're doing when they are authoring, so they won't know which way to use it. 

I'd argue that the test should be better.

**Comment by daviwil** [comment]
It sounds like maybe a better solution here is for `body-complex.json`'s `cookiecuttershark` to be updated to add a property to the type since the check is only looking for `object` schemas with no properties and a single `allOf` reference.  I would hope that there aren't any real specs out there which define "empty" subtypes of a single type as it wouldn't really serve a purpose.

**Comment by MiYanni** [comment]
Could there be a situation where, a service made something like `cookiecuttershark` just because they wanted to use the same schema with a new name? Like, changing the old name would be a breaking change, so they made a new type and made the new APIs use that type instead.

**Comment by daviwil** [comment]
I'm going to make the call to disable `fixUpSchemasThatUseAllOfInsteadOfJustRef` for now to give us more time to discuss how to discourage (or understand the purpose of) that pattern in Swagger specs.  Th ehighest priority right now is to make sure we don't regress behavior of Modelerfour while enabling the use of `lenient-model-deduplication` to unblock partners.

I'm tracking the discussion to re-enable this fixup in issue #278.

**Review by fearthecowboy — APPROVED** [review]

**Comment by daviwil** [comment]
Ended up re-enabling `fixUpSchemasThatUseAllOfInsteadOfJustRef` partially, it now only removes anonymous schemas that follow this pattern and merely warns on schemas that have an actual name.  This should provide the desired behavior without affecting schemas that rely on it.  Thanks @lmazuel and @fearthecowboy for the input!

**Pull request merged**